### DISCUSSION TITLE
What a wonderful world ...

### DISCUSSION THREAD

**Discussion by dorpxam** [AkaneTendo25/musubi-tuner#97] [Ideas]
Since days I was busy with development of some stuff around LTX2 and your wonderful fork. Latest ComfyUI update broke at lot of my (subgraph-based) workflows and many users are going crazy with these radical changes. All this to say that I've fallen behind in keeping up with the progress of your work. The LTX2 documentation is becoming fantastic, and the features are countless. But it's also very difficult for a newbie like me to navigate all the options.

## Self-Flow
My question concerns Self-Flow (Self-Supervised Flow Matching). It seems like a very good feature, and if I understand correctly, it's an option more or less similar to CREPA? Specifically, with the temporal consistency option?

I started running training on the 2.3 model again, specifically on a small test dataset I had set up (sequences from Michael Jackson's "Smooth Criminal" music video). My first attempt was motivated by the desire to see how LTX performed with dancers in symbiosis (synchronized choreography). But I believe that without any features (basic training) and especially with the quality of my source (dataset), the model literally goes haywire (artifacts / robotic movement / loss of quality details).

https://github.com/user-attachments/assets/a1d3d6f4-24fe-44ad-b2c8-b63861186c60

I tried to rework the downscale with Topaz to avoid artifacts and improve the few sequences that serve as a dataset, but overall the result is not satisfactory.

I believe that the error factors are multiple:

- Mid-quality dataset.
- Maybe not enough video (15 in two resolutions)
- Probably a bad learning rate
- And ofcourse, needs some training enhancer, especially in temporal way

The problem isn't with your work, of course. But I need to learn how to use the settings to improve this type of training.

## LoRA Benchmark

To validate that training 2.3 on my machine is not a problem, I launched a quick little training session on a dataset of a female model and even if it is not optimum, the training works very well, even with basic conditions (Adam, 1e-4, no extra features)

Due to Comfy's issues, I had to rewrite my custom nodes to be backward compatible (v3 > v1) and work in both Node 2.0 and legacy modes. Because of this, I wanted to publish my work to share it, but I've fallen behind schedule.

However, I coded a custom benchmark node for LoRa trained with your fork. By pointing directly to the training's "output" directory, I automatically automate the Comfy queue and generate the videos in a dedicated directory. At the end of the generation process, a grid of videos is automatically generated under similar conditions (seed, resolution, etc.).

<img width="2560" height="1392" alt="Image" src="https://github.com/user-attachments/assets/ee6f8bb3-2277-4fa7-be7a-9ebc98c4e4cc" />

The result is obtained in a dedicated directory which allows for multiple tests organized in the following format:

`[ComfyUI-output-folder] / ltx2_lora_benchmark / your_lora_name / 0000N / ... (the produced videos + merged_grid.mp4)`

I still have a few minor improvements to make to simplify the usage parameters. But the idea is very simple. The "Benchmark" node has two outputs: `lora_path` and `save_path`. I've adapted the LoRa loader node to accept a full path (`lora_path`), and `save_path` connects to Comfy's standard "Save Video" function. The rest is automatic. A "`RUN BENCHMARK`" button, and that's it. The result looks something like this (sorry, the video is too heavy, still image here):

<img width="1760" height="640" alt="Image" src="https://github.com/user-attachments/assets/0c525a97-9f6b-41ba-8e15-3c1ea5a9e6a2" />

## Captioning and other stuff

As I mentioned in the discussion section, I'm working on a multi-expert captioner (audio/image/motion/coordinator/captioner) that surpasses anything currently known for a GPU consumer. The mechanism works; however, since the backend is in C++ and depends on chatllm.cpp, I'm reliant on an upcoming update to finalize the project. Nevertheless, the code is already very robust and functional, with a dynamic dashboard like this:

<img width="1115" height="628" alt="Image" src="https://github.com/user-attachments/assets/0e379c24-2609-4a87-969b-7eefe65b3de6" />

I hope to be able to share the project as soon as the overall pipeline is finalized. This project is highly integrated with your fork and dedicated to the community. It is robust enough to allow captioning of both small (LoRA) and very large (Fine-Tuning) datasets.

My guide (FDG) for LTX2 is still relevant. The code is solid, but I need to adapt my nodes for v3 compatibility. Technically, the guide is complete and optimized (as much as possible). I'm currently benchmarking the parameters to fine-tune the user settings (best range/default sweet spot).

Hard-Limit Benchmark:

<img width="1760" height="3840" alt="Image" src="https://github.com/user-attachments/assets/24564940-e704-4cfd-b760-f7cbf79ade56" />

We can clearly see the benefit of setting default ranges and parameters.

## Conclusion:

Regarding my "Smooth Criminal" dataset, I relaunched a 2.3 training run with the following parameters:

```python
--mixed_precision bf16 ^
--max_data_loader_n_workers 4 ^
--persistent_data_loader_workers ^
--dataset_manifest %MANIFEST_PATH% ^
--ltx2_first_frame_conditioning_p 0.1 ^
--ltx2_checkpoint %LTX_23_PATH% ^
--ltx_version 2.3 ^
--ltx_version_check_mode error ^
--ltx2_mode av ^
--separate_audio_buckets ^
--flash_attn ^
--fp8_base ^
--fp8_scaled ^
--blocks_to_swap 32 ^
--use_pinned_memory_for_block_swap ^
--gradient_checkpointing ^
--gradient_accumulation_steps 1 ^
--learning_rate 1e-4 ^
--optimizer_type schedulefree.AdamWScheduleFree ^
--optimizer_args "weight_decay=0.001" "betas=(0.9,0.95)" ^
--network_module networks.lora_ltx2 ^
--network_dim 32 ^
--network_alpha 32 ^
--self_flow ^
--self_flow_args lambda_self_flow=0.1 lambda_audio=0.1 temporal_mode=hybrid lambda_temporal=0.1 lambda_delta=0.05 num_neighbors=2 temporal_granularity=frame ^
--num_timestep_buckets 5 ^
--timestep_sampling shifted_logit_normal ^
--save_every_n_epochs 1 ^
--max_train_epochs 128 ^
--output_name %OUTPUT_NAME% ^
--output_dir %OUTPUT_DIRECTORY% ^
--logging_dir %LOGGING_DIRECTORY% ^
--log_with wandb ^
--log_tracker_name %OUTPUT_NAME%
```

To be honest, I'm currently in "blind" mode, testing the `--self_flow` mode without really understanding it. The training seems to be going well, but we'll have to see if it at least improves consistency in terms of movement, regardless of any remaining artifacts.

<img width="2477" height="626" alt="Image" src="https://github.com/user-attachments/assets/28a1d341-f8eb-4675-aab9-ca1146f85349" />

My datasets for this training are :

```toml
[general]
batch_size = 1
enable_bucket = true
bucket_no_upscale = true

[[datasets]]
resolution = [384, 288]
video_jsonl_file = "E:/Training-2.3/LTX-Smooth/dataset/dataset-288-121.jsonl"
cache_directory = "E:/Training-2.3/LTX-Smooth/dataset/cache-288-121"
frame_extraction = "head"
target_frames = [121]
max_frames = 1441
source_fps = 24.0
target_fps = 24.0
num_repeats = 1

[[datasets]]
resolution = [384, 288]
video_jsonl_file = "E:/Training-2.3/LTX-Smooth/dataset/dataset-288-97.jsonl"
cache_directory = "E:/Training-2.3/LTX-Smooth/dataset/cache-288-97"
frame_extraction = "head"
target_frames = [97]
max_frames = 1441
source_fps = 24.0
target_fps = 24.0
num_repeats = 1

[[datasets]]
resolution = [768, 576]
video_jsonl_file = "E:/Training-2.3/LTX-Smooth/dataset/dataset-576.jsonl"
cache_directory = "E:/Training-2.3/LTX-Smooth/dataset/cache-576"
frame_extraction = "chunk"
target_frames = [25]
max_frames = 1441
source_fps = 24.0
target_fps = 24.0
num_repeats = 1
```

<img width="1665" height="1103" alt="Image" src="https://github.com/user-attachments/assets/99b7d009-9a09-4130-8e6f-ab0d0ab63996" />

The training content is :

```
- 12 videos [384 x 288] 5 seconds @ 24 FPS [121 frames] head
-  3 videos [384 x 288] 4 seconds @ 24 FPS [97 frames] head
- 15 videos [768 x 576] 1 seconds @ 24 FPS [25 frames] 4 or 5 chunks of 1 seconds
```

At 768 x 576, the image quality is not so bad, but I think the smoky atmosphere, the spotlights, the color palette, and the number of characters in the image are driving LTX2 crazy :)

<img width="768" height="576" alt="Image" src="https://github.com/user-attachments/assets/b1dd6b9b-040b-42aa-a590-4833785db112" />

If anyone has any tips on configuring this dataset for testing, I'd love to hear them!

In any case, a million thanks again to you @AkaneTendo25 ! The work you've done these last three months is incredible. The training has reached a level of stability that's evident in the GPU monitoring (it's like a perfectly calm lake), and it's undoubtedly the best tool for working on LTX-2 training. Hats off to you!

**Comment by AkaneTendo25** [comment]
Thanks a lot for the detailed report, and for the kind words too, really appreciate it.

About Self-Flow, I'm still testing it too, so nothing definitive yet and I am not 100% sure the implementation is fully correct. I wouldn't call it CREPA with temporal consistency, it's architecturally quite different. CREPA is a lightweight spatial regularizer. Self-Flow is a full teacher-student framework: it aligns representations against the pretrained base (or an EMA copy), and on top of that it has an motion delta loss that matches temporal differences between student and teacher. So it's a broader regularizer that should help preserve the base model's knowledge across all modalities while you fine-tune. I planned to use it mostly for full-rank fine-tuning, not sure yet how it will behave for LoRA fine-tuning.

Smooth Criminal is perhaps a difficult dataset for the model: not much data, complex choreography, smoke, lights, multiple people, etc. So even with extra tricks, I wouldn't expect miracles there. I'd probably first compare it to a plain training session without Self-Flow. And `1e-4` might be aggressive for this kind of content. Also, one thing about your self_flow_args: you've got `lambda_self_flow=0.1` and `lambda_temporal=0.1` and `lambda_delta=0.05`, a lot of auxiliary losses for small dateset all at once, better start with `lambda_self_flow=0.1` alone, and then add temporal/delta losses if first run doesn't break anything ☺️.  It might be that these params together just add noise that model has to overcome.

(Btw captioner TUI looks great, I'm too in love with TUI-like interfaces like this.)

**Comment by dorpxam** [comment]
You're right, this little dataset is a pain to train. But got the idea. So Self-Flow is probably not my solution :) I'll try with CREPA tonight. Here's the w&b curves, probably do not learn a lot after 1.2/1.4K steps. I'll attempt a lower LR too (2e-5 probably).

<img width="2478" height="624" alt="Image" src="https://github.com/user-attachments/assets/3324699c-efc6-4176-ab6c-ca05a01ed274" />

Even if the video is bad, love the audio. The audio in 2.3 is clearly better, I will try the audio-only training too, the french base is clearly better now (not not north-american accent).

Yes the UI is cool. Using rich: Live(screen=True) you can clearly manage the console like a true UI. Example for the device monitoring:

```python
class DevicePanel(BaseInfoPanel):
    def __init__(self, title: str):
        super().__init__(title=title)
        
    def generate_renderable(self, what: str, ratio_mgb: float, ratio_pct: float, value_mgb: float, value_pct: float):

        progress_grid = Table.grid(expand=True)
        progress_grid.add_column(justify="right")
        progress_grid.add_column(justify="center")
        progress_grid.add_column(justify="left")

        title1 = Text(f"{what}", style="sa")
        title2 = Text("USE", style="sa")

        bar1 = ProgressBar(total=1.0, completed=ratio_mgb, style="sa", complete_style="sc")
        bar2 = ProgressBar(total=1.0, completed=ratio_pct, style="sa", complete_style="sc")

        value1 = Text.assemble((f"{value_mgb:>5.1f} ", "sc"), (f"GB", "sa"))
        value2 = Text.assemble((f"{value_pct:>5.1f} ", "sc"), (f"%", "sa"))

        progress_grid.add_row(Padding(title1, (0, 1, 0, 0)), bar1, Padding(value1, (0, 0, 0, 1)))
        progress_grid.add_row(Padding(title2, (0, 1, 0, 0)), bar2, Padding(value2, (0, 0, 0, 1)))

        return progress_grid


class CPUPanel(DevicePanel):
    def __init__(self):
        super().__init__(title="CPU")
        
    def __rich__(self) -> Panel:
        device = DeviceManager.get_instance()
        usage, used, total = device.get_cpu_datas()

        value_memory_gb = used / (1024**3)
        ratio_memory = (1 / total) * used
        ratio_usage = (1 / 100) * usage

        self.renderable = self.generate_renderable("RAM", ratio_memory, ratio_usage, value_memory_gb, usage)

        return self


class GPUPanel(DevicePanel):
    def __init__(self):
        super().__init__(title="GPU")
        
    def __rich__(self):
        device = DeviceManager.get_instance()
        usage, used, total = device.get_gpu_datas()

        value_memory_gb = used / (1024**3)
        ratio_memory = (1 / total) * used
        ratio_usage = (1 / 100) * usage

        self.renderable = self.generate_renderable("VRAM", ratio_memory, ratio_usage, value_memory_gb, usage)

        return self
```

**Comment by MeiYi-dev** [comment]
> Thanks a lot for the detailed report, and for the kind words too, really appreciate it.
> 
> About Self-Flow, I'm still testing it too, so nothing definitive yet and I am not 100% sure the implementation is fully correct. I wouldn't call it CREPA with temporal consistency, it's architecturally quite different. CREPA is a lightweight spatial regularizer. Self-Flow is a full teacher-student framework: it aligns representations against the pretrained base (or an EMA copy), and on top of that it has an motion delta loss that matches temporal differences between student and teacher. So it's a broader regularizer that should help preserve the base model's knowledge across all modalities while you fine-tune. I planned to use it mostly for full-rank fine-tuning, not sure yet how it will behave for LoRA fine-tuning.
> 
> Smooth Criminal is perhaps a difficult dataset for the model: not much data, complex choreography, smoke, lights, multiple people, etc. So even with extra tricks, I wouldn't expect miracles there. I'd probably first compare it to a plain training session without Self-Flow. And `1e-4` might be aggressive for this kind of content. Also, one thing about your self_flow_args: you've got `lambda_self_flow=0.1` and `lambda_temporal=0.1` and `lambda_delta=0.05`, a lot of auxiliary losses for small dateset all at once, better start with `lambda_self_flow=0.1` alone, and then add temporal/delta losses if first run doesn't break anything ☺️. It might be that these params together just add noise that model has to overcome.
> 
> (Btw captioner TUI looks great, I'm too in love with TUI-like interfaces like this.)

@dorpxam How did the tests on Self-Flow go? Is it worth it to implement?

**Comment by AkaneTendo25** [comment]
@MeiYi-dev I will do a test of my Self-Flow implementation in the coming month, although my case is very specific, it's a full fine-tune of Wan 2.2 on an image-only dataset, and I'm doing a set of ablation studies (for my org) to find the most effective way to do it. I can't say exactly when it will happen, and Self-Flow is just one of the features I want to test, but I'll try to share the quantitative results.

I’m using a different trainer, but the implementation is the same - and by the way, I’m not even sure it is correct xD

upd. Btw here is another [prototype implementation](https://github.com/kohya-ss/musubi-tuner/pull/913) of Self-Flow you did not mention in https://github.com/bghira/SimpleTuner/issues/2632.

**Comment by MeiYi-dev** [comment]
> [@MeiYi-dev](https://github.com/MeiYi-dev) I will do a test of my Self-Flow implementation in the coming month, although my case is very specific, it's a full fine-tune of Wan 2.2 on an image-only dataset, and I'm doing a set of ablation studies (for my org) to find the most effective way to do it. I can't say exactly when it will happen, and Self-Flow is just one of the features I want to test, but I'll try to share the quantitative results.
> 
> I’m using a different trainer, but the implementation is the same - and by the way, I’m not even sure it is correct xD

Thanks for the reply! I am skeptical on Self-Flow as of right now, there are probably better and faster ways to speed up convergence. Currently I think the best way to speed up convergence is:

Using LoKr instead of LoRa
Using TREAD https://arxiv.org/abs/2501.04765
Using CREPA https://arxiv.org/abs/2506.09229
Using Layersync/DINO for alignment https://arxiv.org/abs/2510.12581
Using iREPA https://arxiv.org/abs/2512.10794

All of the above method stacks up nicely and complement each other.
